### PR TITLE
Big count support for datatypes

### PR DESCRIPTION
**Pull request by devreal**
Utilize the count and disp arrays for type-punning of int and MPI_Count arguments to datatype creation functions. Store and pack integer or size_t depending on what is needed. Adjust places where bigcount support for datatypes was missing.

This is work in progress and not yet properly tested, other than compiling through it. I wanted to put it out to get some eyes on it to see whether people agree with the design and that it's the right direction, before putting more effort into testing.

### CODE REVIEW

**Comment by hppritcha** [comment]
okay this is pretty big!  i will probably not be able to get to most of this till Monday.

**Comment by devreal** [comment]
CI currently fails for NVIDIA because hcoll was not adapted but I figure I'll just wait for #13467 to land. Also, I plan to port the datatype tests for bigcount, ideally without having to duplicate the whole code.

**Review by hppritcha — COMMENTED** [review]
why did you remove this?
I'll just need to put it back in for the ABI support.

**Comment by hppritcha** [comment]
> why did you remove this? I'll just need to put it back in for the ABI support.

type_get_envelop.c.in?

**Comment by devreal** [comment]
>> why did you remove this? I'll just need to put it back in for the ABI support.

> type_get_envelop.c.in?

It didn't seem like it was being compiled (it had broken code and no compiled complained). The bigcount version has an additional parameter (number of counts) so it cannot just be generated.

**Comment by hppritcha** [comment]
ah yes.  i forgot in the ABI work i enhanced the python to handle the case of no-big-count vs big-count and could use the template.  so no worries for now.

**Comment by janjust** [comment]
> CI currently fails for NVIDIA because hcoll was not adapted but I figure I'll just wait for #13467 to land. Also, I plan to port the datatype tests for bigcount, ideally without having to duplicate the whole code.

@devreal done, PR merged

**Comment by hppritcha** [comment]
@devreal let me know when you'd like me to review.

**Comment by devreal** [comment]
There are still some issues in the mpi4ai test with 4.1 enabled that I need to resolve (for some of the combiner types). I'm not sure when I'll get to that, likely not before the first week of December.

**Comment by devreal** [comment]
I should say though: the design won't change anymore, so I'd like to get @bosilca's input on this approach.

**Comment by devreal** [comment]
All tests are passing now. I squashed some of the fixups. @hppritcha @bosilca please review

File: ompi/mpi/c/type_get_contents.c [hunk 1]
```diff
@@ -65,9 +65,11 @@ int MPI_Type_get_contents(MPI_Datatype mtype,
         }
     }
 
-    rc = ompi_datatype_get_args( mtype, 1, &max_integers, array_of_integers,
-                            &max_addresses, array_of_addresses,
-                            &max_datatypes, array_of_datatypes, NULL );
+    size_t ci, cl, ca, cd;
```
Comment (hppritcha) [inline ompi/mpi/c/type_get_contents.c:69]: i think you need to do a similar initizlization for max_integers, etc. as in the big count type_get_contents_c.c

**Review by hppritcha — CHANGES_REQUESTED** [review]

**Comment by hppritcha** [comment]
i've not done a full review but found a few items trying to integrate this into the ABI PR #13280

**Comment by hppritcha** [comment]
bingo!  now after a little bit of rebasing fixes I can get #13280 to pass mpi4py declaring OMPI is  MPI 5.0 compliant and not needing to exclude the big count tests! 
I'll do a follow-up review on this PR early next week but this looks promising.

File: ompi/mpi/c/type_get_contents.c (hunk 1 above)
Comment (devreal) [reply ompi/mpi/c/type_get_contents.c:69]: Good catch, not sure how I missed that. I also added checks that the provided sizes are big enough that we don't have large-count in the datatype.

File: ompi/datatype/ompi_datatype_args.c [hunk 2]
```diff
@@ -41,19 +42,22 @@ static inline int
...
     opal_atomic_int32_t ref_count;
     int32_t            create_type;
     size_t             total_pack_size;
-    int32_t            ci;
-    int32_t            ca;
-    int32_t            cd;
-    int*               i;
-    ptrdiff_t* a;
-    ompi_datatype_t**  d;
+    /* TODO: should they be size_t? */
```
Comment (hppritcha) [inline ompi/datatype/ompi_datatype_args.c]: can this todo be removed?

File: ompi/mpi/c/type_contiguous.c.in [hunk 3]
```diff
@@ -56,26 +51,18 @@ PROTOTYPE ERROR_CLASS type_contiguous(COUNT count,
...
-#if OMPI_BIGCOUNT_SRC
-        int icount = (int)count;
-        const int* a_i[1] = {&icount};
-#else
-        const int* a_i[1] = {&count};
-#endif
-        ompi_datatype_set_args( *newtype, 1, a_i, 0, NULL, 1, &oldtype, MPI_COMBINER_CONTIGUOUS );
+        ompi_count_array_t a_i[1] = {OMPI_COUNT_ARRAY_CREATE(&count)};
+        ompi_datatype_set_args( *newtype,
+                                (sizeof(count) == sizeof(size_t)) ? 0 : 1,
```
Comment (hppritcha) [inline ompi/mpi/c/type_contiguous.c.in]: do you mean to use COUNT here?

Comment (hppritcha) [reply ompi/mpi/c/type_contiguous.c.in]: never mind ignore this comment.

**Comment by hppritcha** [comment]
@edgargabriel you may want to review this as it touches multiple ompio files.

**Review by edgargabriel — DISMISSED** [review]
I reviewed the ompio side of the PR, that looks good to me

**Review by hppritcha — DISMISSED** [review]

**Comment by hppritcha** [comment]
are we ready to merge this?

**Comment by devreal** [comment]
I was hoping for a review by @bosilca. Otherwise we're good to go from my pov.

**Comment by janjust** [comment]
@bosilca 
please add your comments

File: ompi/datatype/ompi_datatype_args.c [hunk 4]
```diff
@@ -644,8 +717,11 @@ static ompi_datatype_t* __ompi_datatype_create_from_packed_description( void** p
 
 #if OPAL_ENABLE_HETEROGENEOUS_SUPPORT
     if (need_swap) {
-        for (i = 0 ; i < number_of_length ; ++i) {
-            array_of_length[i] = opal_swap_bytes4(array_of_length[i]);
+        for (i = 0 ; i < number_of_ints ; ++i) {
+            number_of_ints[i] = opal_swap_bytes8(number_of_ints[i]);
```
Comment (janjust) [inline ompi/datatype/ompi_datatype_args.c]: @devreal 

```c
        /* wrong variable name 'number_of_ints[i]' (a scalar) 
         * instead of 'array_of_ints[i]' (the array). Also was using wrong byte 
         * swap size - int is 4 bytes, not 8. */
        for (i = 0 ; i < number_of_ints ; ++i) {
            array_of_ints[i] = opal_swap_bytes4(array_of_ints[i]);
        }
        /* BUGFIX: Was using opal_swap_bytes4 for size_t/MPI_Count which is 8 bytes 
         * on 64-bit systems. Should use opal_swap_bytes8. */
        for (i = 0 ; i < number_of_counts ; ++i) {
            array_of_counts[i] = opal_swap_bytes8(array_of_counts[i]);
        }
```

File: ompi/datatype/ompi_datatype_args.c [hunk 5]
```diff
@@ -684,81 +763,219 @@ static ompi_datatype_t* __ompi_datatype_create_from_args( int32_t* i, MPI_Aint*
...
+            if (l == NULL) {
+                count = i[0];
+                a_i[0] = OMPI_COUNT_ARRAY_CREATE(i);
+                a_i[1] = OMPI_COUNT_ARRAY_CREATE(i + 1);
+                ci = 2 * count + 1;
+            } else {
+                count = l[0];
+                a_i[0] = OMPI_COUNT_ARRAY_CREATE(l);
+                a_i[1] = OMPI_COUNT_ARRAY_CREATE(l + 1);
+                cl = count + 1;
```
Comment (janjust) [inline ompi/datatype/ompi_datatype_args.c]:
```c
cl = 2 * count + 1;
```

File: opal/util/count_disp_array.h [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+#include <stddef.h>
+#include <stdint.h>
+#include <stdbool.h>
+#include "opal_config.h"
+
+/*
+ * NOTE: This code chooses between 64-bit and 32-bit pointers by using the
+ *       least significant bit as a flag (which is possible since these
+ *       pointers will always be multiples of 4 or 8).
+ */
```
Comment (janjust) [inline opal/util/count_disp_array.h:23]: minor  but identified during greptile code review:

**Concerns:**
1. The comment says "multiples of 4 or 8" but only checks bit 0
2. Some architectures might have different alignment requirements
3. User-provided arrays might not always be properly aligned
4. No runtime validation that the assumption holds

**Recommendations:**
1. Add compile-time assertions to verify alignment assumptions:
   ```c
   _Static_assert(alignof(int) >= 2, "int alignment assumption violated");
   _Static_assert(alignof(size_t) >= 2, "size_t alignment assumption violated");
   ```
2. Add runtime checks in debug builds to catch alignment violations
3. Document the alignment requirements clearly in the API

File: ompi/datatype/ompi_datatype_args.c [hunk 7]
```diff
@@ -684,81 +763,219 @@ static ompi_datatype_t* __ompi_datatype_create_from_args( int32_t* i, MPI_Aint*
...
+    }
         /******************************************************************/
     case MPI_COMBINER_HVECTOR_INTEGER:
     case MPI_COMBINER_HVECTOR:
-        ompi_datatype_create_hvector( i[0], i[1], a[0], d[0], &datatype );
         {
-            const int* a_i[2] = {&i[0], &i[1]};
-            ompi_datatype_set_args( datatype, 2, a_i, 1, a, 1, d, MPI_COMBINER_HVECTOR );
+            size_t blocklength;
+            opal_count_array_t a_i[2];
```
Comment (janjust) [inline ompi/datatype/ompi_datatype_args.c]: ptrdiff_t stride;
            /* Need 3 array elements to store count, blocklength, and stride.
             * In bigcount mode, stride is stored in the large count array at position l[2],
             * not in the displacement array 'a'. */
            opal_count_array_t a_i[3];
            size_t ca_val;

File: ompi/datatype/ompi_datatype_args.c [hunk 8]
```diff
@@ -684,81 +763,219 @@ static ompi_datatype_t* __ompi_datatype_create_from_args( int32_t* i, MPI_Aint*
...
+            if (l == NULL) {
+                count          = i[0];
```
Comment (janjust) [inline ompi/datatype/ompi_datatype_args.c:817]: /* Non-bigcount mode: stride comes from displacement array */
                count          = i[0];
                blocklength    = i[1];
                stride         = a[0];

File: ompi/datatype/ompi_datatype_args.c [hunk 9]
```diff
@@ -684,81 +763,219 @@ static ompi_datatype_t* __ompi_datatype_create_from_args( int32_t* i, MPI_Aint*
...
+                blocklength    = i[1];
+                a_i[0] = OMPI_COUNT_ARRAY_CREATE(i);
+                a_i[1] = OMPI_COUNT_ARRAY_CREATE(i + 1);
+                ci = 2;
+            } else { // large count variant
```
Comment (janjust) [inline ompi/datatype/ompi_datatype_args.c:824]: cl = 0;
                ca_val = 1;
            } else {
                /* BUGFIX: Bigcount mode - stride is in large count array at l[2].
                 * Storage format: l[0]=count, l[1]=blocklength, l[2]=stride
                 * Must set cl=3 to match storage, and ca_val=0 since displacement array is unused. */

File: ompi/datatype/ompi_datatype_args.c [hunk 10]
```diff
@@ -684,81 +763,219 @@ static ompi_datatype_t* __ompi_datatype_create_from_args( int32_t* i, MPI_Aint*
...
+                count          = l[0];
+                blocklength    = l[1];
+                a_i[0] = OMPI_COUNT_ARRAY_CREATE(l);
```
Comment (janjust) [inline ompi/datatype/ompi_datatype_args.c:828]: `                stride         = (ptrdiff_t)l[2];`

File: ompi/datatype/ompi_datatype_args.c [hunk 11]
```diff
@@ -684,81 +763,219 @@ static ompi_datatype_t* __ompi_datatype_create_from_args( int32_t* i, MPI_Aint*
...
+                a_i[1] = OMPI_COUNT_ARRAY_CREATE(l + 1);
+                cl = 2;
+            }
+            ompi_datatype_create_hvector( count, blocklength, a[0], d[0], &datatype );
```
Comment (janjust) [inline ompi/datatype/ompi_datatype_args.c]: ompi_datatype_create_hvector( count, blocklength, stride, d[0], &datatype );
            ompi_datatype_set_args( datatype, ci, cl, a_i, ca_val,
                                    (ca_val > 0 ? disp_array : OMPI_DISP_ARRAY_NULL),
                                    1, d, MPI_COMBINER_HVECTOR );

File: ompi/datatype/ompi_datatype_args.c [hunk 12]
```diff
@@ -684,81 +763,219 @@ static ompi_datatype_t* __ompi_datatype_create_from_args( int32_t* i, MPI_Aint*
...
         }
         break;
         /******************************************************************/
     case MPI_COMBINER_HINDEXED_INTEGER:
     case MPI_COMBINER_HINDEXED:
-        ompi_datatype_create_hindexed( i[0], &(i[1]), a, d[0], &datatype );
         {
-            const int* a_i[2] = {&i[0], &i[1]};
-            ompi_datatype_set_args( datatype, i[0] + 1, a_i, i[0], a, 1, d, MPI_COMBINER_HINDEXED );
+            opal_count_array_t a_i[2];
```
Comment (janjust) [inline ompi/datatype/ompi_datatype_args.c]:
```
    case MPI_COMBINER_HINDEXED:
        {
            /* BUGFIX: Need 3 array elements for bigcount mode to store count, blocklengths, and displacements.
             * In bigcount mode, displacements are in large count array at l[1+count], not in displacement array 'a'. */
            opal_count_array_t a_i[3];
            ompi_disp_array_t hindexed_disp;
            size_t ca_val;
            
            if (l == NULL) {
                /* Non-bigcount mode: displacements come from displacement array */
                count = i[0];
                a_i[0] = OMPI_COUNT_ARRAY_CREATE(i);
                a_i[1] = OMPI_COUNT_ARRAY_CREATE(i + 1);
                ci = count + 1;
                cl = 0;
                ca_val = count;
                hindexed_disp = disp_array;
            } else {
                /* BUGFIX: Bigcount mode - displacements are in large count array.
                 * Storage format: l[0]=count, l[1..count]=blocklengths, l[1+count..1+2*count]=displacements
                 * Must set cl=2*count+1 to match storage, and ca_val=0 since displacement array is unused. */
                count = l[0];
                a_i[0] = OMPI_COUNT_ARRAY_CREATE(l);
                a_i[1] = OMPI_COUNT_ARRAY_CREATE(l + 1);
                a_i[2] = OMPI_COUNT_ARRAY_CREATE(l + 1 + count);
                ci = 0;
                cl = 2 * count + 1;
                ca_val = 0;
                hindexed_disp = OMPI_DISP_ARRAY_CREATE(l + 1 + count);
            }
            ompi_datatype_create_hindexed( count, a_i[1], hindexed_disp, d[0], &datatype );
            ompi_datatype_set_args( datatype, ci, cl, a_i, ca_val,
                                    (ca_val > 0 ? hindexed_disp : OMPI_DISP_ARRAY_NULL),
                                    1, d, MPI_COMBINER_HINDEXED );
        }
```

File: ompi/datatype/ompi_datatype_args.c [hunk 13]
```diff
@@ -684,81 +763,219 @@ static ompi_datatype_t* __ompi_datatype_create_from_args( int32_t* i, MPI_Aint*
...
         }
         break;
         /******************************************************************/
     case MPI_COMBINER_STRUCT_INTEGER:
     case MPI_COMBINER_STRUCT:
-        ompi_datatype_create_struct( i[0], &(i[1]), a, d, &datatype );
         {
-            const int* a_i[2] = {&i[0], &i[1]};
-            ompi_datatype_set_args( datatype, i[0] + 1, a_i, i[0], a, i[0], d, MPI_COMBINER_STRUCT );
+            opal_count_array_t a_i[2];
```
Comment (janjust) [inline ompi/datatype/ompi_datatype_args.c]:
```
case MPI_COMBINER_STRUCT:
        {
            /* BUGFIX: Need 3 array elements for bigcount mode to store count, blocklengths, and displacements.
             * In bigcount mode, displacements are in large count array at l[1+count], not in displacement array 'a'. */
            opal_count_array_t a_i[3];
            ompi_disp_array_t struct_disp;
            size_t ca_val;
            
            if (l == NULL) {
                /* Non-bigcount mode: displacements come from displacement array */
                count = i[0];
                a_i[0] = OMPI_COUNT_ARRAY_CREATE(i);
                a_i[1] = OMPI_COUNT_ARRAY_CREATE(i + 1);
                ci = 2 * count + 1;
                cl = 0;
                ca_val = count;
                struct_disp = disp_array;
            } else {
                /* BUGFIX: Bigcount mode - displacements are in large count array.
                 * Storage format: l[0]=count, l[1..count]=blocklengths, l[1+count..1+2*count]=displacements
                 * Must set cl=2*count+1 to match storage, and ca_val=0 since displacement array is unused. */
                count = l[0];
                a_i[0] = OMPI_COUNT_ARRAY_CREATE(l);
                a_i[1] = OMPI_COUNT_ARRAY_CREATE(l + 1);
                a_i[2] = OMPI_COUNT_ARRAY_CREATE(l + 1 + count);
                ci = 0;
                cl = 2 * count + 1;
                ca_val = 0;
                struct_disp = OMPI_DISP_ARRAY_CREATE(l + 1 + count);
            }
            ompi_datatype_create_struct( count, a_i[1], struct_disp, d, &datatype );
            ompi_datatype_set_args( datatype, ci, cl, a_i, ca_val,
                                    (ca_val > 0 ? struct_disp : OMPI_DISP_ARRAY_NULL),
                                    count, d, MPI_COMBINER_STRUCT );

```

File: ompi/datatype/ompi_datatype_args.c [hunk 14]
```diff
@@ -775,14 +992,24 @@ static ompi_datatype_t* __ompi_datatype_create_from_args( int32_t* i, MPI_Aint*
         /******************************************************************/
     case MPI_COMBINER_RESIZED:
         ompi_datatype_create_resized(d[0], a[0], a[1], &datatype);
-        ompi_datatype_set_args( datatype, 0, NULL, 2, a, 1, d, MPI_COMBINER_RESIZED );
+        ompi_datatype_set_args( datatype, 0, 0, NULL, 2, disp_array, 1, d, MPI_COMBINER_RESIZED );
         break;
         /******************************************************************/
     case MPI_COMBINER_HINDEXED_BLOCK:
-        ompi_datatype_create_hindexed_block( i[0], i[1], a, d[0], &datatype );
         {
```
Comment (janjust) [inline ompi/datatype/ompi_datatype_args.c:1025]:
```
case MPI_COMBINER_HINDEXED_BLOCK:
        {
            size_t bLength = 0;
            /* BUGFIX: Need 3 array elements for bigcount mode to store count, blocklength, and displacements.
             * In bigcount mode, displacements are in large count array at l[2], not in displacement array 'a'. */
            opal_count_array_t a_i[3];
            ompi_disp_array_t hib_disp;
            size_t ca_val;
            
            if (l == NULL) {
                /* Non-bigcount mode: displacements come from displacement array */
                count = i[0];
                bLength = i[1];
                a_i[0] = OMPI_COUNT_ARRAY_CREATE(&count);
                a_i[1] = OMPI_COUNT_ARRAY_CREATE(&bLength);
                ci = 2;
                cl = 0;
                ca_val = count;
                hib_disp = disp_array;
            } else {
                /* BUGFIX: Bigcount mode - displacements are in large count array.
                 * Storage format: l[0]=count, l[1]=blocklength, l[2..1+count]=displacements
                 * Must set cl=count+2 to match storage, and ca_val=0 since displacement array is unused. */
                count = l[0];
                bLength = l[1];
                a_i[0] = OMPI_COUNT_ARRAY_CREATE(&count);
                a_i[1] = OMPI_COUNT_ARRAY_CREATE(&bLength);
                a_i[2] = OMPI_COUNT_ARRAY_CREATE(l + 2);
                ci = 0;
                cl = count + 2;
                ca_val = 0;
                hib_disp = OMPI_DISP_ARRAY_CREATE(l + 2);
            }
            ompi_datatype_create_hindexed_block( count, bLength, hib_disp, d[0], &datatype );
            ompi_datatype_set_args( datatype, ci, cl, a_i, ca_val,
                                    (ca_val > 0 ? hib_disp : OMPI_DISP_ARRAY_NULL),
                                    1, d, MPI_COMBINER_HINDEXED_BLOCK );
        }
```

**Review by janjust — CHANGES_REQUESTED** [review]
@devreal I think I found a few issues, maybe, I ran the code through a greptile.

https://github.com/janjust/ompi/commits/main-PR13460/

I made one commit with suggested changes here - not sure if relevant, please take a look.

**Comment by janjust** [comment]
@devreal which tests did you use to test this PR, if any?

**Comment by devreal** [comment]
Thanks @janjust! I will address them in the coming days. The test I used was mpi4py with the MPI version in OMPI bumped to 5 (in the `VERSION` file).

**Comment by devreal** [comment]
Not sure why the ASAN mpi4py fails. I cannot reproduce it on my system. Will dig into it.

**Comment by janjust** [comment]
@devreal @bosilca 

I think we got it working check out my PR here: it adds 3 small changes to this PR:
https://github.com/janjust/ompi/commits/main-PR13460/

```
$  mpirun -np 1 ./tests/ompi-tests/mpich_tester/install/bin/large_count
PASSED
tomislavj@funk11 /hpc/mtr_scrap/users/tomislavj/hpcx-dev
$ mpirun -np 1 ./tests/ompi-tests/mpich_tester/install/bin/large_type
PASSED
tomislavj@funk11 /hpc/mtr_scrap/users/tomislavj/hpcx-dev
$ mpirun -np 2 -H funk11:1,funk12:1 -x LD_LIBRARY_PATH ./tests/ompi-tests/mpich_tester/install/bin/large_type_sendrec
PASSED
PASSED
tomislavj@funk11 /hpc/mtr_scrap/users/tomislavj/hpcx-dev
$ mpirun -np 1 ./tests/ompi-tests/mpich_tester/install/bin/large_count
PASSED
```

```
tomislavj@NV-F54Q6M3:~/tomislavj/hpcx-stack-dev/ompi$ git log -n 4 --oneline
3a5aa0f4a2 (HEAD -> main-PR13460, origin/main-PR13460) Fix MPI_Get_elements_c clipping at INT_MAX instead of SSIZE_MAX
a7ebc51bb5 Fix same regression in MPI_Status_set_elements_x for derived datatypes
c8877a3600 Fix MPI_Status_set_elements regression for derived datatypes
c3d8fb2bb1 Address review comments: fixes and documentation
tomislavj@NV-F54Q6M3:~/tomislavj/hpcx-stack-dev/ompi$ git diff HEAD~3 HEAD
diff --git a/ompi/mpi/c/get_elements.c.in b/ompi/mpi/c/get_elements.c.in
index 49483f92f5..4949777396 100644
--- a/ompi/mpi/c/get_elements.c.in
+++ b/ompi/mpi/c/get_elements.c.in
@@ -66,10 +66,13 @@ PROTOTYPE ERROR_CLASS Get_elements(STATUS status, DATATYPE datatype, COUNT_OUT c
 
     ret = ompi_datatype_get_elements (datatype, status->_ucount, &internal_count);
     if (OMPI_SUCCESS == ret || OMPI_ERR_VALUE_OUT_OF_BOUNDS == ret) {
-        if (OMPI_SUCCESS == ret && internal_count <= INT_MAX) {
+        /* Clip at the maximum value representable in the output type:
+         * INT_MAX for int (regular variant), SSIZE_MAX for MPI_Count (_c variant). */
+        size_t count_max = (sizeof(*count) > sizeof(int)) ? (size_t) SSIZE_MAX : (size_t) INT_MAX;
+        if (OMPI_SUCCESS == ret && internal_count <= count_max) {
             *count = internal_count;
         } else {
-            /* If we have more elements that we can represent with a signed int then we must
+            /* If we have more elements than we can represent then we must
              * set count to MPI_UNDEFINED (MPI 3.0).
              */
             *count = MPI_UNDEFINED;
diff --git a/ompi/mpi/c/status_set_elements.c.in b/ompi/mpi/c/status_set_elements.c.in
index 953a7aef7d..3dd09c7878 100644
--- a/ompi/mpi/c/status_set_elements.c.in
+++ b/ompi/mpi/c/status_set_elements.c.in
@@ -65,12 +65,7 @@ PROTOTYPE ERROR_CLASS status_set_elements(STATUS_OUT status, DATATYPE datatype,
         return MPI_SUCCESS;
     }
 
-    if( ompi_datatype_is_predefined(datatype) ) {
-        ompi_datatype_type_size( datatype, &size );
-        status->_ucount = count * size;
-    } else {
-        ompi_datatype_set_element_count( datatype, count, &size );
-        status->_ucount = size;
-    }
+    ompi_datatype_type_size( datatype, &size );
+    status->_ucount = (size_t)count * size;
     return MPI_SUCCESS;
 }
diff --git a/ompi/mpi/c/status_set_elements_x.c.in b/ompi/mpi/c/status_set_elements_x.c.in
index c24b24f9c5..b7d5af84d7 100644
--- a/ompi/mpi/c/status_set_elements_x.c.in
+++ b/ompi/mpi/c/status_set_elements_x.c.in
@@ -65,12 +65,7 @@ PROTOTYPE ERROR_CLASS status_set_elements_x(STATUS_OUT status, DATATYPE datatype
         return MPI_SUCCESS;
     }
 
-    if( ompi_datatype_is_predefined(datatype) ) {
-        ompi_datatype_type_size( datatype, &size );
-        status->_ucount = count * size;
-    } else {
-        ompi_datatype_set_element_count( datatype, count, &size );
-        status->_ucount = size;
-    }
+    ompi_datatype_type_size( datatype, &size );
+    status->_ucount = (size_t)count * size;
     return MPI_SUCCESS;
 }
```

**Comment by devreal** [comment]
@janjust I'm trying to understand why the change in https://github.com/janjust/ompi/commit/c8877a36008c233e4cf18d4ecf5fd6613cc26091 is needed. There is no documentation for `opal_datatype_set_element_count` but my understanding is that for derived types we may not just rely on the type size. Maybe @bosilca can take a look at this?

**Comment by devreal** [comment]
@janjust I think I got it, with a fix to the MPICH tester and a fix to `MPI_Get_elements_c`. Thanks to @bosilca for explaining how the datatype engine works :)

I will squash everything down before merge.

File: ompi/datatype/ompi_datatype_args.c [hunk 15]
```diff
@@ -644,8 +716,11 @@ static ompi_datatype_t* __ompi_datatype_create_from_packed_description( void** p
 
 #if OPAL_ENABLE_HETEROGENEOUS_SUPPORT
     if (need_swap) {
-        for (i = 0 ; i < number_of_length ; ++i) {
-            array_of_length[i] = opal_swap_bytes4(array_of_length[i]);
+        for (i = 0 ; i < number_of_ints ; ++i) {
```
Comment (bosilca) [inline ompi/datatype/ompi_datatype_args.c:719]: Aren't these two swapped ? array_of_ints I would say should be using `opal_swap_bytes4` while array_of_counts should go with `opal_swap_bytes8` ?

File: ompi/datatype/ompi_datatype_args.c [hunk 16]
```diff
@@ -335,17 +390,18 @@ int32_t ompi_datatype_get_args( const ompi_datatype_t* pData, int32_t which,
     switch(which){
     case 0:     /* GET THE LENGTHS */
         *ci = pArgs->ci;
+        *cl = pArgs->cl;
         *ca = pArgs->ca;
         *cd = pArgs->cd;
         *type = pArgs->create_type;
         break;
     case 1:     /* GET THE ARGUMENTS */
-        if(*ci < pArgs->ci || *ca < pArgs->ca || *cd < pArgs->cd) {
```
Comment (bosilca) [inline ompi/datatype/ompi_datatype_args.c:343]: Why was the check removed ?

File: ompi/mpi/c/type_contiguous.c.in [hunk 17]
```diff
@@ -56,26 +52,18 @@ PROTOTYPE ERROR_CLASS type_contiguous(COUNT count,
...
```
Comment (bosilca) [inline ompi/mpi/c/type_contiguous.c.in:64]: This is ugly and unreadable. Why not using OMPI_BIGCOUNT_SRC instead (as in other places) ?

File: ompi/datatype/ompi_datatype_args.c [hunk 18]
```diff
@@ -117,92 +142,113 @@ int32_t ompi_datatype_set_args( ompi_datatype_t* pData,
...
         break;
 
     case MPI_COMBINER_HINDEXED_BLOCK:
-        pArgs->i[0] = i[0][0];
-        pArgs->i[1] = i[1][0];
+        copy_count_array(1, &pi, &pl, counts[0]);
+        copy_count_array(1, &pi, &pl, counts[1]);
+        if (cl > 0) {
+            // copy the displacements
+            size_t count = opal_count_array_get(counts[0], 0);
```
Comment (bosilca) [inline ompi/datatype/ompi_datatype_args.c:248]: These are displacements so they are supposed to be MPI_Aint and not MPI_Count (especially for the H function).
Also, are you sure the arguments should be `counts[0]` and not `counts[2]`?

File: ompi/mpi/c/type_create_hindexed.c.in [hunk 19]
```diff
@@ -73,50 +66,37 @@ PROTOTYPE ERROR_CLASS type_create_hindexed(COUNT count,
...
-        iarray_of_blocklengths[i] = (int)array_of_blocklengths[i];
-        iarray_of_displacements[i] = (MPI_Aint)array_of_displacements[i];
-    }
-#else
-    iarray_of_blocklengths = (int *)array_of_blocklengths;
-    iarray_of_displacements = (MPI_Aint *)array_of_displacements;
-#endif
 
-    rc = ompi_datatype_create_hindexed( count, iarray_of_blocklengths, iarray_of_displacements,
+    /*
```
Comment (bosilca) [inline ompi/mpi/c/type_create_hindexed.c.in:71]: Why would the displacement be of type `MPI_Count` ? Isn't this an oversight with the MPI standard ?

File: VERSION [hunk 20]
```diff
@@ -20,7 +20,7 @@ minor=1
 release=0
 
 # MPI Standard Compliance Level
-mpi_standard_version=3
+mpi_standard_version=4
```
Comment (bosilca) [inline VERSION:23]: Isn't this premature ?

File: opal/mca/btl/uct/btl_uct_component.c [hunk 21]
```diff
@@ -20,6 +20,7 @@
  * Copyright (c) 2019-2025 Google, LLC. All rights reserved.
  * Copyright (c) 2019      Intel, Inc.  All rights reserved.
  * Copyright (c) 2022      IBM Corporation.  All rights reserved.
+ * Copyright (c) 2026      Stony Brook University.  All rights reserved.
```
Comment (bosilca) [inline opal/mca/btl/uct/btl_uct_component.c:23]: why ? There are several files where only the copyright has been added. Please fix.

File: opal/util/count_disp_array.h [hunk 22]
```diff
@@ -0,0 +1,242 @@
...
+    if (OPAL_LIKELY(array & 0x1L)){
+        int *iptr = (int *)(array & ~0x1L);
+        iptr[i] = (int)val;
+    } else {
+        size_t *sptr = (size_t *)array;
+        sptr[i] = val;
+    }
+}
+
+typedef intptr_t opal_disp_array_t;
```
Comment (bosilca) [inline opal/util/count_disp_array.h:138]: I have an issue with this second type, mostly because I think it can be folded into the first one. 

However, I see a special case for this because of the hindex type creation where the displacement can be either int or MPI_Count, while on all the other type creation functions it can be either int or MPI_Aint. If we merge this into a single array type then we have 3 choices for the displacement array: int, MPI_Count or MPI_Aint.

**Review by bosilca — COMMENTED** [review]
This is a huge PR, with a lot of drastic changes all over the code. I think no amount of code review will be able to properly vet everything, which means we need to rely on testing to cover all cases. Thus my biggest question is: do we have nay test coverage for these new capabilities ?

File: ompi/mpi/c/type_create_hindexed.c.in (hunk 19 above)
Comment (devreal) [reply ompi/mpi/c/type_create_hindexed.c.in:71]: Unfortunately, it is `MPI_Aint` in the non-bigcount API and `MPI_Count` in the bigcount API. I think that change was made consistently (since `MPI_Count` must be big enough to hold `MPI_Aint`)

File: opal/mca/btl/uct/btl_uct_component.c (hunk 21 above)
Comment (devreal) [reply opal/mca/btl/uct/btl_uct_component.c:23]: Will fix that. Somehow this came in through the update-copyright script

File: VERSION (hunk 20 above)
Comment (devreal) [reply VERSION:23]: This is the last thing missing for 4.0 compliance.

File: ompi/datatype/ompi_datatype_args.c (hunk 15 above)
Comment (devreal) [reply ompi/datatype/ompi_datatype_args.c:719]: Will fix.

File: ompi/datatype/ompi_datatype_args.c (hunk 18 above)
Comment (devreal) [reply ompi/datatype/ompi_datatype_args.c:248]: The caller passes the displacement as MPI_Count in the large-counts, instead of the displacement array.

File: ompi/datatype/ompi_datatype_args.c (hunk 16 above)
Comment (devreal) [reply ompi/datatype/ompi_datatype_args.c:343]: Will bring that baack and add check for `cl`.

File: ompi/mpi/c/type_contiguous.c.in (hunk 17 above)
Comment (devreal) [reply ompi/mpi/c/type_contiguous.c.in:64]: Will change that to separate calls.

File: opal/mca/accelerator/cuda/accelerator_cuda.c [hunk 23]
```diff
@@ -10,6 +10,7 @@
  * Copyright (c) 2024      The University of Tennessee and The University
  *                         of Tennessee Research Foundation.  All rights
  *                         reserved.
+ * Copyright (c) 2026      Stony Brook University.  All rights reserved.
```
Comment (devreal) [inline opal/mca/accelerator/cuda/accelerator_cuda.c:13]: Will remove these

File: ompi/datatype/ompi_datatype_args.c (hunk 18 above)
Comment (devreal) [reply ompi/datatype/ompi_datatype_args.c:248]: TODO: adapt the MPICH test to test the combiner content: https://github.com/open-mpi/ompi-tests/blob/master/mpich_tester/mpich_datatype/large_type.c

**Comment by janjust** [comment]
need to finish the combiner test, but it's basically ready to go

**Comment by janjust** [comment]
After some discussion with the team - we're going to merge, retest and fix post merge if anything.
@hppritcha @devreal @edgargabriel  will cherry pick to v6.0.x
@bosilca fyi

**Review by janjust — APPROVED** [review]
